<a href="https://colab.research.google.com/github/coderzaman/Python-For-ML/blob/main/Python_for_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

program 01: Write a program to display the messages “Hello”, “World” and “Good Bye”. Each of the three messages should get displayed on a different line.

In [ ]:
print("Hello")
print("World")
print("Good Bye")

Hello
World
Good Bye


Program 02: Write a basic program to make use of the end key and display the messages “Hello” “World” and “Good Bye” in one line


**using a special argument named end=''.**

In [ ]:
print("Hello ", end='')
print("World ", end='')
print("Good Bye", end='')

Hello World Good Bye

# Multiple assignments

In [ ]:
p, q = 20, 30
print(p, q)

20 30


**swap number**

In [ ]:
a = 10
b = 30

a,b = b,a

print(a,b)

30 10


Program 03: Write a simple program to calculate the area of a rectangle in Python

formula: area = lenght * breadth

In [ ]:
length = int(input("Please Enter Length: "))
breadth = int(input("Please Enter Breadth: "))

area = length * breadth

print("Area of Rectangle is", area)

Please Enter Length: 5
Please Enter Breadth: 4
Area of Rectangle is 20


**eval() Function**

* The eval() function in Python takes a string as input and evaluates it as a Python expression.

* Normally, when we use input(), Python always takes user input as a string, even if the user enters a number.

* To convert that string into a number, we usually wrap it with int(), float(), etc.


**But if we use eval(), Python automatically interprets the input in the correct type.**

* If the user types a number → it becomes a number.

* If the user types a list → it becomes a list.

* If the user types an expression → it evaluates the expression.


**⚠️ Warning:**
eval() is unsafe if you directly use user input (because it can execute harmful code). It should be avoided in real applications unless you fully trust the input.

In [ ]:
x = eval(input("Enter your input: "))
print(type(x))

x = input("Enter your input: ")
print(type(x))

Enter your input: 12
<class 'int'>
Enter your input: 12
<class 'str'>


Formatting number and strings

In [ ]:
x = 20.345123
print(format(x, '.3f'))
print(type(x))

20.345
<class 'float'>


In [ ]:
# It retruns string 7.3f and print begin a space(Like Think with common sense space in string formatted output should be string)
x = 20.345123
x = format(x, "7.3f")
print(x)
print(type(x))

 20.345
<class 'str'>


In [ ]:
x = 20.345123
x = format(x, "7.3f")
print(x)
print(type(x))

 20.345
<class 'str'>


In [ ]:
x = 20.345123
x = format(x, "<7.3f")
print(x)
print(type(x))

20.345 
<class 'str'>


In [ ]:
x = eval(x)
print(x)
print(type(x))

20.345
<class 'float'>


**'.2%' meaning:**

. → decimal point specification.

2 → number of digits after the decimal.

% → convert the number to percentage (multiplies by 100 and adds % sign).

So:

First, 314.344234 × 100 = 31434.4234

Then, keep 2 digits after decimal → 31434.42

Finally, add the % sign → "31434.42%"

In [ ]:
print(format(314.344234, '.2%'))

31434.42%


In [ ]:
x = 0.5
x = format(x, '.2%')
x

'50.00%'

### 🔹 Common Math Functions in Python (`math` module)

1. **`ceil(x)`**

   * Rounds **x** up to the nearest integer.
   * Example: `ceil(4.2) = 5`

2. **`floor(x)`**

   * Rounds **x** down to the nearest integer (largest integer ≤ x).
   * Example: `floor(4.9) = 4`

3. **`exp(x)`**

   * Returns **eˣ** (where `e = 2.718...`).
   * Example: `exp(1) = 2.718...`

4. **`log(x)`**

   * Returns the **natural logarithm (base e)** of `x`.
   * Example: `log(2.718...) = 1`

5. **`log(x, base)`**

   * Returns the **logarithm of x with the given base**.
   * Example: `log(100, 10) = 2`

6. **`sqrt(x)`**

   * Returns the **square root of x**.
   * Example: `sqrt(25) = 5`

7. **`sin(x)`**

   * Returns the **sine of x**, where `x` is in **radians**.
   * Example: `sin(3.14159/2) = 1`

8. **`asin(x)`**

   * Returns the **inverse sine (arcsine)** of `x`, result is in **radians**.
   * Example: `asin(1) = 1.5708 rad (≈ 90°)`

9. **`cos(x)`**

   * Returns the **cosine of x**, where `x` is in **radians**.
   * Example: `cos(0) = 1`

10. **`acos(x)`**

    * Returns the **inverse cosine (arccos)** of `x`, result is in **radians**.
    * Example: `acos(1) = 0`

11. **`tan(x)`**

    * Returns the **tangent of x**, where `x` is in **radians**.
    * Example: `tan(0) = 0`

12. **`degrees(x)`**

    * Converts angle **x** from **radians → degrees**.
    * Example: `degrees(3.14159) = 180°`

13. **`radians(x)`**

    * Converts angle **x** from **degrees → radians**.
    * Example: `radians(180) = 3.14159 rad`


In [ ]:
import math

In [ ]:
x = math.floor(10.3443)
x

10

In [ ]:
x = 10.00001
x = math.ceil(x)
x

11

In [ ]:
x = 10.573434
x = round(x,2)
x

10.57

In [ ]:
x = 10.573434
x = round(x,0)
x

11.0

In [ ]:
x = math.log(math.e)
x

1.0

In [ ]:
x = math.log(100,10)
x

2.0

In [ ]:
x = math.sqrt(49)
print(x)

7.0


In [ ]:
x = math.sin(math.pi/2)
x

1.0

In [ ]:
x = math.asin(1)
x

1.5707963267948966

In [ ]:
x = math.cos(0)
x

1.0

In [ ]:
x = math.acos(1)
x

0.0

In [ ]:
x = math.tan(0)
x

0.0

In [ ]:
x = math.degrees(math.pi)
x

180.0

In [ ]:
x = math.radians(180)
x

3.141592653589793

# The ord and chr Functions
ord function returns ACII value of Character




In [ ]:
ord('A')

65

In [ ]:
ord('a')

97

retrun char base on ascii value

In [ ]:
chr(97)

'a'

In [ ]:
chr(255)

'ÿ'

## Summary


### 🐍 Python Basics (Easy Notes)

1. **Tokens** → Python breaks each program into small parts called *tokens*.
   (Examples: keywords, variables, operators, etc.)

2. **Literals** → Fixed values written directly in the program.

   * Example: `18` (integer), `21.98` (float), `"Q"` (character), `"Hello"` (string).

3. **`type()`** → Built-in function to check the type of any value.

4. **Keywords** → Special reserved words in Python.

   * Example: `if`, `else`, `while`, `for`, `class`, `True`, `False` etc.
   * Keywords **cannot** be used as variable names.

5. **Identifiers** → Names given to variables, functions, classes, etc.

   * Can contain letters, digits, underscores.
   * Cannot start with a digit.
   * Cannot be a keyword.
   * Example: `Name`, `Roll_NO`, `A1`, `_address`.

6. **Everything in Python is an object.**

7. **`int()`** → Converts a string or number to an integer.

   * Example: `int("10") → 10`

8. **`float()`** → Converts a string or number to a decimal (floating point).

   * Example: `float("12.5") → 12.5`

9. **Boolean (bool)** → Data type with only two values: `True` or `False`.

10. **`print()`** → Displays output on the screen.

11. **`input()`** → Takes input from the user (always as string).

12. **`format()`** → Used to create formatted strings.

13. **Types of Tokens** →

    * **Keywords**
    * **Identifiers / Variables**
    * **Operators**
    * **Delimiters**
    * **Literals**

14. **Operators in Python**

    * Arithmetic: `+  -  *  /  //  %  **`
    * Relational: `==  !=  <  >  <=  >=`
    * Logical: `and  or  not`
    * Bitwise: `&  |  ~  ^  <<  >>`

15. **Delimiters** → Symbols with special meaning.

    * Examples: `( ) [ ] { } , : . = ;`
    * Assignment operators: `+=  -=  *=  /=  //=  %=  &=  |=  ^=  >>=  <<=  **=`


## 🔹 Key Terms in Python

1. **`chr()`**

   * Returns the **character** for a given ASCII (or Unicode) value.
   * Example: `chr(65)` → `'A'`

2. **`ord()`**

   * Returns the **ASCII (or Unicode) value** of a given character.
   * Example: `ord('A')` → `65`

3. **`print()`**

   * Displays the output on the screen.
   * Example: `print("Hello")` → `Hello`

4. **`end` (argument in `print()`)**

   * Used inside the `print()` function to control what comes at the end of the output.
   * By default, `end="\n"` (new line).
   * Example:

     ```python
     print("Hello", end=" ")
     print("World")
     ```

     Output: `Hello World`

5. **`format()`**

   * Used to format strings and numbers.
   * Example:

     ```python
     print("Value: {:.2f}".format(3.14159))
     ```

     Output: `Value: 3.14`

6. **`input()`**

   * Accepts data (always as a string) from the user.
   * Example:

     ```python
     name = input("Enter your name: ")
     ```

7. **`int()`**

   * Converts a value (string or float) into an integer.
   * Example: `int("10")` → `10`

8. **`str()`**

   * Converts a value into a string.
   * Example: `str(123)` → `"123"`

9. **`type()`**

   * Returns the exact data type of a value.
   * Example: `type(10)` → `<class 'int'>`

10. **Identifier**

    * A **name given to variables, functions, or classes**.
    * Example: `x = 5` → here `x` is an identifier.

11. **Tokens**

    * The smallest unit of a Python program.
    * Types of tokens:

      * Keywords (`if`, `while`, `for`)
      * Identifiers (`x`, `total`)
      * Literals (`10`, `"Hello"`)
      * Operators (`+`, `-`, `*`, `/`)
      * Punctuators (`,`, `:`, `()`)

12. **Inbuilt Math Functions** (from the `math` module):

    * `abs(x)` → absolute value
    * `max(a, b, …)` → largest value
    * `round(x, n)` → round to n decimal places
    * `ceil(x)` → round up
    * `floor(x)` → round down
    * `log(x)` → natural log (base e)
    * `exp(x)` → eˣ
    * `sqrt(x)` → square root
    * `sin(x), cos(x), tan(x)` → trigonometric functions (input in radians)
    * `asin(x), acos(x), atan(x)` → inverse trigonometric functions (output in radians)
    * `degrees(x)` → converts radians → degrees
    * `radians(x)` → converts degrees → radians


# Excercise

### 🔹 Rounding Halfway Rules

1. **Rounding Half Up (common “school” rounding):**

   * If the fractional part is exactly 0.5:

     * **Positive numbers:** round **up** to the next greater integer.

       * Example: `4.5 → 5` ✅
     * **Negative numbers:** round **down** to the next more negative integer.

       * Example: `-4.5 → -5` ✅

   > Think: “halfway goes **away from zero**.”



2. **Rounding Half Down:**

   * If the fractional part is exactly 0.5:

     * **Positive numbers:** round **down** to the smaller integer.

       * Example: `4.5 → 4`
     * **Negative numbers:** round **up** (toward zero).

       * Example: `-4.5 → -4`

   > Think: “halfway goes **toward zero**.”


3. **Rounding Half to Even (Python’s `round()`):**

   * If the fractional part is exactly 0.5, round to **the nearest even integer**.

     * Example:

       * `2.5 → 2` (2 is even)
       * `3.5 → 4` (4 is even)
       * `-2.5 → -2`
       * `-3.5 → -4`

---

### 🔹 Summary for Positive vs Negative

| Rule                  | Positive 0.5 | Negative -0.5 |
| --------------------- | ------------ | ------------- |
| Half Up (school)      | 1.5 → 2      | -1.5 → -2     |
| Half Down             | 1.5 → 1      | -1.5 → -1     |
| Half to Even (Python) | 1.5 → 2      | -1.5 → -2     |

* ✅ **Positive value rounding**: halfway goes **toward the greater integer** in “half up”
* ✅ **Negative value rounding**: halfway goes **toward the more negative integer** in “half up”


In [ ]:
round(1.5)

2

In [ ]:
round(-1.5)

-2

In [ ]:
_ = 10
_

10

In [ ]:
a = 15//4 #floor division
a

3

Problem 1:

In [ ]:
length = float(input("Enter the length of the rectangle: "))
width = float(input("Enter the width of the rectangle: "))

if length == width:
    print("The rectangle is a square.")
else:
    print("The rectangle is not a square.")


Enter the length of the rectangle: 2
Enter the width of the rectangle: 2
The rectangle is a square.


Problem 2:

In [ ]:
num1 = float(input("Enter the first number: "))
num2 = float(input("Enter the second number: "))
num3 = float(input("Enter the third number: "))

if num1 >= num2 and num1 >= num3:
    largest = num1
elif num2 >= num3:
    largest = num2
else:
    largest = num3

print("The largest number is:", largest)

Enter the first number: 12
Enter the second number: 4
Enter the third number: -4
The largest number is: 12.0


Problem 3

In [ ]:
number = int(input("Enter a number: "))

if number % 2 == 0:
    print(number, "is even.")
else:
    print(number, "is odd.")

Enter a number: 15
15 is odd.


Proble 4

In [ ]:
score = int(input("Enter the student's score: "))

if score > 90:
  print("A")
elif score > 80:
  print("B")
elif score >= 60:
  print("C")
else:
  print("D")

score = int(input("Enter the student's score: "))

if score > 90:
  print("A")
elif score > 80 and score <=90:
  print("B")
elif score >= 60 and score <= 80:
  print("C")
else:
  print("D")



Enter the student's score: 78
C
Enter the student's score: 56
D


Problem 5

In [ ]:
year = int(input("Enter a year: "))

if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
    print(year, "is a leap year.")
else:
    print(year, "not a leap year.")

Enter a year: 15
15 not a leap year.


In [ ]:
x = range(10)
x

x[9]

9

In [ ]:
x = range(10)

print(f"1. Object created by range(10): {x}")
print(f"2. The type of the object is: {type(x)}")

1. Object created by range(10): range(0, 10)
2. The type of the object is: <class 'range'>


In [ ]:
x = list(range(2,10,2))
x

[2, 4, 6, 8]

In [ ]:
for i in range(1,11):
  for j in range(1,11):
    print(f"{i} * {j} = {i*j}")
  print()
  print()

1 * 1 = 1
1 * 2 = 2
1 * 3 = 3
1 * 4 = 4
1 * 5 = 5
1 * 6 = 6
1 * 7 = 7
1 * 8 = 8
1 * 9 = 9
1 * 10 = 10


2 * 1 = 2
2 * 2 = 4
2 * 3 = 6
2 * 4 = 8
2 * 5 = 10
2 * 6 = 12
2 * 7 = 14
2 * 8 = 16
2 * 9 = 18
2 * 10 = 20


3 * 1 = 3
3 * 2 = 6
3 * 3 = 9
3 * 4 = 12
3 * 5 = 15
3 * 6 = 18
3 * 7 = 21
3 * 8 = 24
3 * 9 = 27
3 * 10 = 30


4 * 1 = 4
4 * 2 = 8
4 * 3 = 12
4 * 4 = 16
4 * 5 = 20
4 * 6 = 24
4 * 7 = 28
4 * 8 = 32
4 * 9 = 36
4 * 10 = 40


5 * 1 = 5
5 * 2 = 10
5 * 3 = 15
5 * 4 = 20
5 * 5 = 25
5 * 6 = 30
5 * 7 = 35
5 * 8 = 40
5 * 9 = 45
5 * 10 = 50


6 * 1 = 6
6 * 2 = 12
6 * 3 = 18
6 * 4 = 24
6 * 5 = 30
6 * 6 = 36
6 * 7 = 42
6 * 8 = 48
6 * 9 = 54
6 * 10 = 60


7 * 1 = 7
7 * 2 = 14
7 * 3 = 21
7 * 4 = 28
7 * 5 = 35
7 * 6 = 42
7 * 7 = 49
7 * 8 = 56
7 * 9 = 63
7 * 10 = 70


8 * 1 = 8
8 * 2 = 16
8 * 3 = 24
8 * 4 = 32
8 * 5 = 40
8 * 6 = 48
8 * 7 = 56
8 * 8 = 64
8 * 9 = 72
8 * 10 = 80


9 * 1 = 9
9 * 2 = 18
9 * 3 = 27
9 * 4 = 36
9 * 5 = 45
9 * 6 = 54
9 * 7 = 63
9 * 8 = 72
9 * 9 = 81
9 * 10 = 90


10 * 1

Pattern 01

In [ ]:
for i in range(10):
  for j in range(i):
    print("* ", end="")
  print()


* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 
* * * * * * * * 
* * * * * * * * * 


Problem 2

In [ ]:
for i in range(10):
  for j in range(10):
    if i <= j:
      print(" ", end="")
    else:
      print("* ", end="")
  print()

          
*          
* *         
* * *        
* * * *       
* * * * *      
* * * * * *     
* * * * * * *    
* * * * * * * *   
* * * * * * * * *  


Problem 3

In [ ]:
bajar_list = [["Alu", "Peyaj", "Roshun"], [10,20,30], ['a','b','c','d','e']]

for bajar in bajar_list:
  for item in bajar:
    print(item, end=" ")
  print()

Alu Peyaj Roshun 
10 20 30 
a b c d e 


Problem 4

In [ ]:
number = int(input("Whose number multiflication you wnat: "))

for i in range(1,11):
  print(f"{number} * {i} = {number*i}")

Whose number multiflication you wnat: 12
12 * 1 = 12
12 * 2 = 24
12 * 3 = 36
12 * 4 = 48
12 * 5 = 60
12 * 6 = 72
12 * 7 = 84
12 * 8 = 96
12 * 9 = 108
12 * 10 = 120


Problem 5

In [ ]:
number = int(input("Whose number factorial you wnat: "))
fact = 1
for i in range(number, 1, -1):
  fact = fact * i

print(f"Factorial of {number} is {fact}")

Whose number factorial you wnat: 12
Factorial of 12 is 479001600


Problem 6

In [ ]:
n = int(input("Enter Length of Fibonacci Number Series: "))
first_number = 0
second_number = 1
print(first_number, end=" ")
print(second_number, end=" ")

fibo = 0

for i in range(n-2):
  fibo = first_number + second_number
  print(fibo, end=" ")
  first_number = second_number
  second_number = fibo




Enter Length of Fibonacci Number Series: 5
0 1 1 2 3 

Problem 7

In [ ]:
numbers = [10,20,30,40,50]
i = 0
for number in numbers:
  i+=1

print(f"Length Of Length is: {i}")
print(f'Length of Length is: {len(numbers)}')

Length Of Length is: 5
Length of Length is: 5


Problem 8

In [ ]:
number = int(input("Enter a number: "))
count = 0

while number != 0:
  number = number // 10
  count += 1

print("Number of digits:", count)


Enter a number: 12
Number of digits: 2


Problem 9

In [ ]:
number = int(input("Enter a number: "))
length = len(str(number))

temp = number

sum = 0

while number != 0:
  digit = number % 10
  sum += digit ** length
  number = number // 10

if temp == sum:
  print(f"{temp} is an armstrong number")
else:
  print(f"{temp} is not an armstrong number")

Enter a number: 153
153 is an armstrong number


In [ ]:
number = input("Enter a number: ")
length = len(number)

sum = 0

for i in number:
  sum += int(i)**length

if sum == int(number):
  print("Armstrong Number")
else:
  print("Not Armstrong Number")

Enter a number: 153
Armstrong Number


Problem 10

In [ ]:
number = int(input("Enter a Number: "))

rev_number = 0

while number != 0:
  digit = number % 10
  rev_number = rev_number * 10 + digit
  number = number // 10

print(f"Reverse Number is: {rev_number}")

Enter a Number: 153
Reverse Number is: 351


# List

In [ ]:
li = [10,29,39,59,[39,32,2], "abc", "def", (10,20,30)]
print(li)

[10, 29, 39, 59, [39, 32, 2], 'abc', 'def', (10, 20, 30)]


In [ ]:
print(li[0:5])

[10, 29, 39, 59, [39, 32, 2]]


In [ ]:
print(li[0:5:2]) #start:end:step

[10, 39, [39, 32, 2]]


In [ ]:
print(li[-1:-3:]) # for negetive indexing we need must step

[]


In [ ]:
print(li[-1:-(len(li)+1):-1])

[(10, 20, 30), 'def', 'abc', [39, 32, 2], 59, 39, 29, 10]


In [ ]:
print(li[-1::-1])

[(10, 20, 30), 'def', 'abc', [39, 32, 2], 59, 39, 29, 10]


In [ ]:
print(li[::-1])

[(10, 20, 30), 'def', 'abc', [39, 32, 2], 59, 39, 29, 10]


In [ ]:
a = "HelloWorld!"
a = list(a)

print(a)

['H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'd', '!']


In [ ]:
print(a.count('o'))

2


In [ ]:
a.pop()
a

['H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l']

In [ ]:
a.append("Hello")
a

['H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'Hello']

In [ ]:
a.insert(0,45)

In [ ]:
a

[45, 'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'Hello']

In [ ]:
a.extend(['a','b','c'])

In [ ]:
a

[45, 'H', 'e', 'l', 'l', 'o', 'W', 'o', 'r', 'l', 'Hello', 'a', 'b', 'c']

In [ ]:
nested_list = [[1, 2, 3], [4, 5], [6, 7, 8]]
flat_list = [item for sublist in nested_list for item in sublist]
print(flat_list)

[1, 2, 3, 4, 5, 6, 7, 8]


In [ ]:
a = input().split()
print(a)

hello world bangladesh
['hello', 'world', 'bangladesh']


In [ ]:
a = list(map(int, input().split()))
a

1 2 3 4 5 6 


[1, 2, 3, 4, 5, 6]

In [ ]:
a = list(map(float, input().split()))
a

5.2 6.7 8.9 5.3


[5.2, 6.7, 8.9, 5.3]

In [ ]:
number = [1,2,3,4,5]

number = [n*5 for n in number]
number

[5, 10, 15, 20, 25]

In [ ]:
character = [l for l in "Hello world"]
character

['H', 'e', 'l', 'l', 'o', ' ', 'w', 'o', 'r', 'l', 'd']

In [ ]:
number = [x for x in range(1,21) if x%3==0]
number

[3, 6, 9, 12, 15, 18]

In [ ]:
number = [x for x in range(1,100) if x%3==0 and x%5==0]
number

[15, 30, 45, 60, 75, 90]

In [ ]:
number = [x for x in range(1,100) if x%3==0 or x%5==0]
number

[3,
 5,
 6,
 9,
 10,
 12,
 15,
 18,
 20,
 21,
 24,
 25,
 27,
 30,
 33,
 35,
 36,
 39,
 40,
 42,
 45,
 48,
 50,
 51,
 54,
 55,
 57,
 60,
 63,
 65,
 66,
 69,
 70,
 72,
 75,
 78,
 80,
 81,
 84,
 85,
 87,
 90,
 93,
 95,
 96,
 99]

In [ ]:
number = [f"{x} is Even" if x%2==0 else f"{x} is Odd" for x in range(1,11)]
number

['1 is Odd',
 '2 is Even',
 '3 is Odd',
 '4 is Even',
 '5 is Odd',
 '6 is Even',
 '7 is Odd',
 '8 is Even',
 '9 is Odd',
 '10 is Even']

In [ ]:
matrix = [[1,2],[3,4],[5,6],[7,8]]
a = []
for i in range(2):
  b = []
  for col in matrix:
    b.append(col[i])
  a.append(b)
a

[[1, 3, 5, 7], [2, 4, 6, 8]]

In [ ]:
matrix = [[1,2],[3,4],[5,6],[7,8],[9,10]]
a = []

for i in range(len(matrix[0])):
  b = []
  for col in matrix:
    b.append(col[i])
  a.append(b)
a

[[1, 3, 5, 7, 9], [2, 4, 6, 8, 10]]

In [ ]:
matrix = [[1,2],[3,4],[5,6],[7,8],[9,10]]

t_matrix = [[col[i] for col in matrix] for i in range(2)]
t_matrix

[[1, 3, 5, 7, 9], [2, 4, 6, 8, 10]]

In [ ]:
matrix = [[1,2],[3,4],[5,6],[7,8],[9,10]]

t_matrix = [[col[i] for col in matrix if col[i]%3!=0] for i in range(2)]
t_matrix

[[1, 5, 7], [2, 4, 8, 10]]

In [ ]:
items = [4,6,4,3,3,4,3,4,3,4]

new_list = []
for item in items:
  if items.count(item) > 3 and item not in new_list:
    new_list.append(item)

print(new_list)


[4, 3]


In [ ]:
#[[1, 2, 3, 4], [0, 2, 3, 4], [0, 1, 3, 4], [0, 1, 2, 4], [0, 1, 2, 3]]

nested_list = [[j for j in range(5) if i!=j] for i in range(5)]
print(nested_list)

[[1, 2, 3, 4], [0, 2, 3, 4], [0, 1, 3, 4], [0, 1, 2, 4], [0, 1, 2, 3]]


In [ ]:
s= [0, "Hello", [], None, [1]]
print([bool(x) for x in s])

[False, True, False, False, True]


In [ ]:
tpl = (1, 2, 3, [4, 5])
tpl[3].append(6)
print(tpl)

(1, 2, 3, [4, 5, 6])


In [ ]:
st = {1, 2, 2, 3, 'A'}
print(st)

{1, 2, 3, 'A'}


In [ ]:
st = {1, 2, 2, 3, 'A'}
print(st)


{1, 2, 3, 'A'}


In [ ]:
tpl = (1, 2, 3, [4, 5])
tpl[3].append(6)
print(tpl)


(1, 2, 3, [4, 5, 6])


In [ ]:
x = 5
y = "5"
print(type(x), type(y))


<class 'int'> <class 'str'>


In [ ]:
print(7 / 2)

print(type(5.))

3.5
<class 'float'>


In [ ]:
num = float(input())
print(type(num))


7
<class 'float'>


In Python, **operator precedence** determines the order in which operations are evaluated.
Here’s the precedence (from highest to lowest) among the given operators:

1. **not** → highest
2. **==**
3. **and**
4. **or** → lowest

So, the order is:

```
not  >  ==  >  and  >  or
```


**Operator precedence** works among
`not`, `==`, `and`, and `or` in Python

### 🔹 Example 1: `not` vs `and`

```python
x = True
y = False

print(not x and y)
```

**Step-by-step:**

* `not` has higher precedence than `and`
* So it becomes: `(not x) and y` → `False and False` → **False**

✅ **Output:** `False`


### 🔹 Example 2: `and` vs `or`

```python
a = True
b = False
c = True

print(a or b and c)
```

**Step-by-step:**

* `and` has higher precedence than `or`
* So it becomes: `a or (b and c)` → `a or False` → **True**

✅ **Output:** `True`

### 🔹 Example 3: `not` vs `==`

```python
x = True
y = False

print(not x == y)
```

**Step-by-step:**

* `not` has higher precedence than `==`
* So it becomes: `(not x) == y` → `False == False` → **True**

✅ **Output:** `True`

### 🔹 Example 4: combining all

```python
print(not True or False and True == False)
```

**Operator precedence order:**

1. `not`
2. `==`
3. `and`
4. `or`

**Step-by-step:**

* `not True` → `False`
* `True == False` → `False`
* `False and False` → `False`
* `False or False` → `False`

✅ **Output:** `False`


In [ ]:
inp = input()
inp = inp.split()
inp = [int(x) for x in inp]

min_n = min(inp)
max_n = max(inp)

print(min_n, max_n)

1 2 3
1 3


In [ ]:
no_of_inp = int(input())
inp = input().split()

even_count = 0
odd_count = 0
postive_count = 0
negateive_count = 0

# Convert the input strings to integers
inp = [int(x) for x in inp]


for i in range(no_of_inp):
  if inp[i] % 2 == 0:
    even_count += 1
  else:
    odd_count += 1

  if inp[i] > 0:
    postive_count += 1
  elif inp[i] < 0 :
    negateive_count += 1

print("Even:", even_count)
print("Odd:", odd_count)
print("Positive:", postive_count)
print("Negative:", negateive_count)

5
-5 0 -3 -4 12
Even: 3
Odd: 2
Positive: 1
Negative: 3


In [ ]:
no_of_inp = int(input())

for i in range(no_of_inp):
  inp = input()
  inp = inp[::-1]

  for c in inp:
    print(c, end=" ")

  print()


2
123
3 2 1 
1200
0 0 2 1 


In [ ]:
x = 5

if x > 3:
  print("A")
elif x > 1:
  print("B")
else:
  print("Kichu e print korbena")

A


In [ ]:
s=0
for i in range(1, 5):
  if i%2 == 0:
    s += 1
  else:
    s -= 1
print(s)

0


In [ ]:
count = 0
for x in [1,2,3,4,5]:
  if x % 2 != 0:
    continue

In [ ]:
w_weight = int(input())

if w_weight % 2 == 0 and w_weight != 2:
  print("YES")
else:
  print("NO")

10
YES


In [ ]:
no_of_inp = int(input())

for i in range(no_of_inp):
  inp = input()
  inp_len = len(inp)
  if inp_len <= 10:
    print(inp)
  else:
    cal = inp_len-2
    print(inp[0] + str(cal)+inp[-1])

4
word
word
localization
l10n
internationalization
i18n
pneumonoultramicroscopicsilicovolcanoconiosis
p43s


In [ ]:
no_of_inp = int(input())
count = 0
for i in range(no_of_inp):
  inp = input().split()

  if inp.count("1") >= 2:
    count +=1

print(count)

2
1 0 
0 1 1
1


In [ ]:
no_of_inp = int(input())
count = 0

for i in range(no_of_inp):
  inp = input()

  if  '++' in inp:
    count +=1
  else:
    count -=1

print(count)



2
X++
--X
0


In [ ]:
import math
inp = [int (x) for x in input().split()]

need_stone_for_cover_width = math.ceil(inp[0]/inp[2])
need_stone_for_cover_length = math.ceil(inp[1]/inp[2])

print(need_stone_for_cover_width*need_stone_for_cover_length)


6 6 4
4


In [ ]:
no_of_inp = list(map(int, input().split()))

score_position = (no_of_inp[1])-1
number_of_players = no_of_inp[0]
inp = list(map(int, input().split()))

score = inp[score_position]
count_qualified_player  = 0

for i in range(number_of_players):
  if inp[i] >= score and inp[i] > 0:
    count_qualified_player +=1

print(count_qualified_player)


5 5
3 2 1 0 0
3


In [ ]:
import math
inp = list(map(int, input().split()))

area_of_domino = inp[0] * inp[1]
print(math.floor(area_of_domino/2))

3 3
4


In [ ]:
matrix_input = list(map(int, input().split()))

matrix = []
matrix_index = 0

track_i = -1
track_j = -1

for i in range(0,5):
  row = []
  for j in range(0,5):
    row.append(matrix_input[matrix_index])
    if matrix_input[matrix_index] == 1:
      track_i = i
      track_j = j
    matrix_index += 1
  matrix.append(row)

print(abs(2-track_i) + abs(2-track_j))



0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 00 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0
1


In [ ]:
word1 = input().lower()
word2 = input().lower()

if word1 > word2:
  print(1)
elif word1 < word2:
  print(-1)
else:
  print(0)

abcdefg
AbCdEfF
1


In [ ]:
tree_collected_fruits = list(map(int, input().split()))

fruits_input_per_tree = list(map(int, input().split()))
fruits_input_per_tree.sort(reverse=True)

toatal_tree = tree_collected_fruits[0]
total_fruits = tree_collected_fruits[1]

tree_count = toatl_fruit = 0
success = 0
for i in range(toatal_tree):
  toatl_fruit += fruits_input_per_tree[i]
  tree_count += 1
  if toatl_fruit >= total_fruits:
    success = 1
    break

if success != 1:
  print("Shombhob na")
else:
  print(tree_count)

4 15
1 1 2 14
2


In [ ]:
no_inp = int(input())
inp = list(input().split())
max = 0

for i in range(no_inp):
  highest_duplicate = inp.count(inp[i])
  if highest_duplicate > max:
    max = highest_duplicate

print(max)

5
2 2 5 5 6
2


In [ ]:
word = "Pythonista"
print(word[-4])

i


In [ ]:
word = "HelloWorld"
print(word[2:8:2])

loo


In [ ]:
word = "banana bandana"
print(word.count("ana"))

2


In [ ]:
word = "banana bandana"
print(word.count("ana"))

2


In [ ]:
msg = "red,green,blue"
print("-".join(msg.split(",")[::-1]))

blue-green-red


In [ ]:
num = [2,4,6,8,10]
num[2] = num[2] * num[0]
print(num)

[2, 4, 12, 8, 10]


In [ ]:
letter = ['a','b','c','d','e']
print(letter[1:4][::-2])

['d', 'b']


In [ ]:
stack = []

for i in range(1,4):
  stack.append(i)

stack.pop()
print(stack[-1])

2


In [ ]:
quee = [10,20,30,40]
quee.pop(1)
quee.append(50)
quee.pop(0)
print(quee[0])

30


In [ ]:
number = [1,2,3,4]

square = [x**2 for x in number if x%2==0]
print(square)

[4, 16]


In [ ]:
x = [1,2,3,2,4]
x.remove(4)
x.pop()
print(x)

[1, 2, 3]


In [ ]:
no_of_inp = int(input())
inp = list(map(int, input().split()))
minimum = min(inp)
print(minimum, inp.index(minimum)+1)

5
5 5 3 2 1
1 5


In [ ]:
no_of_inp = int(input())
inp = list(map(int, input().split()))
minimum = min(inp)
maximum = max(inp)

index_of_minimum = inp.index(minimum)
index_of_maximum = inp.index(maximum)

inp[index_of_minimum], inp[index_of_maximum] = inp[index_of_maximum], inp[index_of_minimum]

for i in inp:
  print(i, end=" ")

5
4 1 3 10 8
4 10 3 1 8 

In [ ]:
no_of_inp = int(input())
inp = input()
sum = 0
for i in inp:
  sum+=int(i)

print(sum)


5
13305
12


In [ ]:
no_of_inp = int(input())

for i in range(no_of_inp):
  inp = input()

  if "010" in inp or "101" in inp:
    print("Good")
  else:
    print("Bad")


2
11111110
Bad
10101010101010
Good


In [ ]:
inp = input().upper()
inp = inp.replace("EGYPT", " ")
print(inp)

ITALYKOREAEGYPTEGYPTALGERIAEGYPTZ
ITALYKOREA  ALGERIA Z


In [ ]:
inp = list(map(float, input().split()))
brightness = inp[0]
threshold = inp[1]

if brightness >= threshold:
  print("ON")
else:
  print("OFF")

40 50
OFF


In [ ]:
inp = list(map(float, input().split()))
x = inp[0]
min_v = inp[1]
max_v = inp[2]

norm = (x - min_v) / (max_v - min_v)

print(f"{norm:.2f}")

2.5 0 10
0.25


In [ ]:
no_of_inp = int(input())

sum_of_loss = 0
taraget = 0
for i in range(no_of_inp+1):
  inp = float(input())

  if i == 0:
    target = inp;
  else:
    sum_of_loss += inp


average_loss = sum_of_loss / no_of_inp
if average_loss <= target:
  print("PASS")
else:
  print("RETRY")


5
0.40
0.42
0.45
0.39
0.41
0.44
RETRY


In [ ]:
no_of_inp = int(input())

yes = 0
no = 0

for i in range(no_of_inp):
  inp = input().upper()
  if inp == "YES":
    yes += 1
  else:
    no += 1

if yes >= no:
  print("ACCEPT")
else:
  print("REJECT")

2
NO
yes
ACCEPT


In [ ]:
inp = input().lower()

if "happy" in inp or "joy" in inp or "smile" in inp:
  print("Happy Mood")
elif "sad" in inp or "cry" in inp or "angry" in inp:
  print("Sad Mood")
else:
  print("Neutral Mood")

i am angry today
Sad Mood


In [ ]:
inp = list(map(int, input().split()))

total_brightness = 0

for i in inp:
  total_brightness += i

average_brightness = total_brightness / len(inp)

if average_brightness < 85:
  print("Dark Image")
elif average_brightness <= 170:
  print("Normal Image")
else:
  print("Bright Image")

10 20 30
Dark Image


In [ ]:
inp = input().upper().split()
average_of_A = inp.count("A") * 100 / len(inp)
average_of_B = inp.count("B") * 100 / len(inp)

if average_of_A > 70 or average_of_B > 70:
  print("Biased Model")
else:
  print("Fair Model")


A A  B B B
Fair Model


In [ ]:
#ai, data, model, learn, train, neural
inp = input().lower().split()

ai_count = inp.count("ai") + inp.count("data") + inp.count("model") + inp.count("learn") + inp.count("train") + inp.count("neural")

if ai_count >= 2:
  print("AI Detected")
else:
  print("Not AI Related")


i love to train a neural model
AI Detected
